File: macos/Sources/SnapikCore/Exporting/FileExportService.swift
import Foundation

/// Port of `src/Snapik.Infrastructure/Exporting/FileExportService.cs`.
///
/// The C# renderer writes PNG bytes into a caller-provided `Stream`; the Swift
/// `ExportImageRendering` protocol instead returns `Data` for the rendered PNG (simpler surface
/// for a Foundation-only, no-CoreImage Core target — the Mac app target owns the actual pixel
/// rendering). This service still validates the PNG signature, hashes it, and writes it to disk
/// itself, preserving the "no partial revision on invalid renderer output" guarantee.
public final class FileExportService: ExportService {
    private static let pngSignature: [UInt8] = [137, 80, 78, 71, 13, 10, 26, 10]

    private let renderer: ExportImageRendering
    private let timeProvider: TimeProvider
    /// The letter a package of one capture keeps instead of the "A" its position would give it: the
    /// card it was copied from shows that letter, and the picture, the file name and the text have
    /// to agree with it. `nil`, and a package numbers itself by position, byte for byte as before.
    private let singleCaptureLabel: String?

    public init(
        renderer: ExportImageRendering,
        timeProvider: TimeProvider = SystemTimeProvider(),
        singleCaptureLabel: String? = nil
    ) {
        self.renderer = renderer
        self.timeProvider = timeProvider
        self.singleCaptureLabel = singleCaptureLabel
    }

    public func prepare(session: SnapikSession, sessionDirectory: URL) async throws -> PreparedExport {
        try SessionValidation.validate(session)
        guard !session.captures.isEmpty else {
            throw SnapikError.invalidOperation("Cannot export a session without captures.")
        }

        let sessionRoot = sessionDirectory.standardizedFileURL
        let exportId = SBGuid()
        let exportsRoot = sessionRoot.appendingPathComponent("exports", isDirectory: true)
        let staging = exportsRoot.appendingPathComponent(".staging-\(exportId.digitsLowercase)", isDirectory: true)
        let revisionSuffix = String(format: "%06d", session.revision)
        let finalDirectory = exportsRoot.appendingPathComponent(
            "revision-\(revisionSuffix)-\(exportId.digitsLowercase)", isDirectory: true)
        try FileManager.default.createDirectory(at: staging, withIntermediateDirectories: true)

        do {
            var images: [ExportImageEntry] = []
            for (index, capture) in session.captures.enumerated() {
                let label: String
                if let only = singleCaptureLabel, session.captures.count == 1 {
                    label = only
                } else {
                    label = try CaptureLabels.forIndex(index)
                }
                // Port of `FileExportService.cs:43-45`: the user opens these files in a folder
                // of their own, and "01-A.png" sorts and reads like a page number. The guid of the
                // capture stays in `manifest.images[].captureId`, and nothing takes the name apart
                // again.
                let fileName = String(format: "%02d", index + 1) + "-\(label).png"
                let imagePath = staging.appendingPathComponent(fileName)
                let sourceImagePath = try Self.resolveSessionPath(
                    sessionRoot: sessionRoot, relativePath: capture.sourceImagePath)

                guard FileManager.default.fileExists(atPath: sourceImagePath.path) else {
                    throw SnapikError.fileNotFound(
                        "Capture source image does not exist: \(sourceImagePath.path)")
                }

                let context = ExportImageContext(displayLabel: label, captureIndex: index, sourceImagePath: sourceImagePath)
                let rendered = try await renderer.renderPNG(
                    capture: capture, annotations: capture.annotations, context: context)

                try Self.validatePNG(rendered)
                try Self.writeNew(rendered, to: imagePath)

                images.append(
                    ExportImageEntry(
                        captureId: capture.id,
                        displayLabel: label,
                        fileName: fileName,
                        sha256: SHA256.hexString(rendered),
                        byteLength: Int64(rendered.count)))
            }

            // Port of `FileExportService.cs:63-73`: captures without notes produce no text at
            // all. The package is then images only, and an empty prompt.md would just be an empty
            // file for the user to open; the manifest carries empty strings, and a reader of it has
            // to allow for them.
            let promptText = try PromptGenerator(singleCaptureLabel: singleCaptureLabel).generate(session)
            let promptFileName = promptText.isEmpty ? "" : "prompt.md"
            var promptSha256 = ""
            if !promptText.isEmpty {
                let promptData = Data(promptText.utf8)
                try Self.writeNew(promptData, to: staging.appendingPathComponent(promptFileName))
                promptSha256 = SHA256.hexString(promptData)
            }

            let manifest = ExportManifest(
                exportId: exportId,
                sessionId: session.id,
                revision: session.revision,
                createdAtUtc: timeProvider.utcNow(),
                images: images,
                promptFileName: promptFileName,
                promptSha256: promptSha256,
                promptText: promptText,
                captureCount: session.captures.count,
                noteCount: Self.countNotes(session))

            let manifestPath = staging.appendingPathComponent("manifest.json")
            let manifestData = try SnapikJson.encoder.encode(manifest)
            try Self.writeNew(manifestData, to: manifestPath)

            try FileManager.default.moveItem(at: staging, to: finalDirectory)
            return PreparedExport(rootDirectory: finalDirectory, manifest: manifest)
        } catch {
            if FileManager.default.fileExists(atPath: staging.path) {
                try? FileManager.default.removeItem(at: staging)
            }
            throw error
        }
    }

    private static func writeNew(_ data: Data, to url: URL) throws {
        try data.write(to: url, options: .atomic)
    }

    private static func countNotes(_ session: SnapikSession) -> Int {
        var count = ExportText.hasContent(session.globalNote) ? 1 : 0
        for capture in session.captures {
            count += ExportText.hasContent(capture.note) ? 1 : 0
            count += capture.annotations.filter { ExportText.hasContent($0.note) }.count
        }
        return count
    }

    private static func resolveSessionPath(sessionRoot: URL, relativePath: String) throws -> URL {
        let resolved = sessionRoot.appendingPathComponent(relativePath).standardizedFileURL
        var rootPrefix = sessionRoot.standardizedFileURL.path
        if !rootPrefix.hasSuffix("/") {
            rootPrefix += "/"
        }
        guard resolved.path.lowercased().hasPrefix(rootPrefix.lowercased()) else {
            throw SnapikError.invalidData("Capture image path escapes the session directory.")
        }
        return resolved
    }

    private static func validatePNG(_ data: Data) throws {
        guard data.count >= pngSignature.count, Array(data.prefix(pngSignature.count)) == pngSignature else {
            throw SnapikError.invalidData("The export renderer did not produce a PNG image.")
        }
    }
}

File: src/Snapik.Infrastructure/Exporting/FileExportService.cs
using System.Collections.Immutable;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using Snapik.Core.Exporting;
using Snapik.Core.Models;
using Snapik.Infrastructure.Serialization;

namespace Snapik.Infrastructure.Exporting;

public sealed class FileExportService(IExportImageRenderer renderer, TimeProvider? timeProvider = null) : IExportService
{
    private static readonly byte[] PngSignature = [137, 80, 78, 71, 13, 10, 26, 10];
    private readonly TimeProvider _timeProvider = timeProvider ?? TimeProvider.System;

    public async Task<PreparedExport> PrepareAsync(
        SnapikSession session,
        string sessionDirectory,
        CancellationToken cancellationToken = default)
    {
        SessionValidation.Validate(session);
        if (session.Captures.IsEmpty)
        {
            throw new InvalidOperationException("Cannot export a session without captures.");
        }

        ArgumentException.ThrowIfNullOrWhiteSpace(sessionDirectory);
        var sessionRoot = Path.GetFullPath(sessionDirectory);
        var exportId = Guid.NewGuid();
        var exportsRoot = Path.Combine(sessionRoot, "exports");
        var staging = Path.Combine(exportsRoot, $".staging-{exportId:N}");
        var finalDirectory = Path.Combine(exportsRoot, $"revision-{session.Revision:D6}-{exportId:N}");
        Directory.CreateDirectory(staging);

        try
        {
            var images = ImmutableArray.CreateBuilder<ExportImageEntry>(session.Captures.Length);
            for (var index = 0; index < session.Captures.Length; index++)
            {
                cancellationToken.ThrowIfCancellationRequested();
                var capture = session.Captures[index];
                var label = CaptureLabels.ForIndex(index);
                // The user opens these files in a folder of their own: "01-A.png" sorts and reads
                // like a page number. The guid of the capture stays in manifest.images[].captureId.
                var fileName = $"{index + 1:D2}-{label}.png";
                var imagePath = Path.Combine(staging, fileName);
                var sourceImagePath = ResolveSessionPath(sessionRoot, capture.SourceImagePath);
                if (!File.Exists(sourceImagePath))
                {
                    throw new FileNotFoundException("Capture source image does not exist.", sourceImagePath);
                }

                await using (var output = new FileStream(imagePath, FileMode.CreateNew, FileAccess.Write, FileShare.None, 65536, FileOptions.Asynchronous))
                {
                    await renderer.RenderAsync(capture, new ExportImageContext(label, index, sourceImagePath), output, cancellationToken);
                }

                await ValidatePngAsync(imagePath, cancellationToken);
                var imageInfo = new FileInfo(imagePath);
                images.Add(new ExportImageEntry(capture.Id, label, fileName, await Sha256Async(imagePath, cancellationToken), imageInfo.Length));
            }

            // Captures without notes produce no text at all: the package is then images only, and
            // an empty prompt.md would just be an empty file for the user to open.
            var promptText = new PromptGenerator().Generate(session);
            var promptFileName = promptText.Length == 0 ? string.Empty : "prompt.md";
            var promptSha256 = string.Empty;
            if (promptText.Length > 0)
            {
                var promptPath = Path.Combine(staging, promptFileName);
                await File.WriteAllTextAsync(promptPath, promptText, new UTF8Encoding(false), cancellationToken);
                promptSha256 = await Sha256Async(promptPath, cancellationToken);
            }

            var manifest = new ExportManifest(
                exportId,
                session.Id,
                session.Revision,
                _timeProvider.GetUtcNow(),
                images.MoveToImmutable(),
                promptFileName,
                promptSha256,
                promptText,
                session.Captures.Length,
                CountNotes(session));

            var manifestPath = Path.Combine(staging, "manifest.json");
            await using (var manifestStream = new FileStream(manifestPath, FileMode.CreateNew, FileAccess.Write, FileShare.None, 65536, FileOptions.Asynchronous))
            {
                await JsonSerializer.SerializeAsync(manifestStream, manifest, SnapikJson.Options, cancellationToken);
            }

            Directory.Move(staging, finalDirectory);
            return new PreparedExport(finalDirectory, manifest);
        }
        catch
        {
            if (Directory.Exists(staging))
            {
                Directory.Delete(staging, true);
            }

            throw;
        }
    }

    private static int CountNotes(SnapikSession session) =>
        (ExportText.HasContent(session.GlobalNote) ? 1 : 0) +
        session.Captures.Sum(capture =>
            (ExportText.HasContent(capture.Note) ? 1 : 0) +
            capture.Annotations.Count(annotation => ExportText.HasContent(annotation.Note)));

    private static string ResolveSessionPath(string sessionRoot, string relativePath)
    {
        var resolved = Path.GetFullPath(Path.Combine(sessionRoot, relativePath));
        var rootPrefix = Path.TrimEndingDirectorySeparator(sessionRoot) + Path.DirectorySeparatorChar;
        if (!resolved.StartsWith(rootPrefix, StringComparison.OrdinalIgnoreCase))
        {
            throw new InvalidDataException("Capture image path escapes the session directory.");
        }

        return resolved;
    }

    private static async Task ValidatePngAsync(string path, CancellationToken cancellationToken)
    {
        var signature = new byte[PngSignature.Length];
        await using var stream = File.OpenRead(path);
        var bytesRead = await stream.ReadAsync(signature, cancellationToken);
        if (bytesRead != signature.Length || !signature.SequenceEqual(PngSignature))
        {
            throw new InvalidDataException("The export renderer did not produce a PNG image.");
        }
    }

    private static async Task<string> Sha256Async(string path, CancellationToken cancellationToken)
    {
        await using var stream = File.OpenRead(path);
        var hash = await SHA256.HashDataAsync(stream, cancellationToken);
        return Convert.ToHexStringLower(hash);
    }
}
